src/SqlBinding/SqlAsyncCollector.cs

```diff
@@ -90,55 +92,284 @@ public async Task FlushAsync(CancellationToken cancellationToken = default)
+            /// </summary>
+            public IDictionary<string, string> Columns { get; }
+
+            /// <summary>
+            /// List of strings containing each column and its type. ex: ["Cost int", "LastChangeDate datetime(7)"]
+            /// </summary>
+            public IEnumerable<string> ColumnDefinitions => Columns.Select(c => $"{c.Key} {c.Value}");
+
+            /// <summary>
+            /// T-SQL merge statement generated from primary keys
+            /// and column names for a specific table.
+            /// </summary>
+            public string MergeQuery { get; }
+
+            public TableInformation(IEnumerable<MemberInfo> primaryKeys, IDictionary<string, string> columns, string mergeQuery)
+            {
+                this.PrimaryKeys = primaryKeys;
+                this.Columns = columns;
+                this.MergeQuery = mergeQuery;
+            }
+
+            /// <summary>
+            /// Generates SQL query that can be used to retrieve the Primary Keys of a table
+            /// </summary>
+            public static string GetPrimaryKeysQuery(string schema, string tableName) 
+            {
+                return $@"
+                    select 
+                        {ColumnName}
+                    from
+                        INFORMATION_SCHEMA.TABLE_CONSTRAINTS tc
+                    inner join
+                        INFORMATION_SCHEMA.CONSTRAINT_COLUMN_USAGE ccu on ccu.CONSTRAINT_NAME = tc.CONSTRAINT_NAME AND ccu.TABLE_NAME = tc.TABLE_NAME
+                    where
+                        tc.CONSTRAINT_TYPE = 'PRIMARY KEY'
+                    and
+                        tc.TABLE_NAME = '{tableName}'
+                    and
+                        tc.TABLE_SCHEMA = {schema}";
+            }
+
+            /// <summary>
+            /// Generates SQL query that can be used to retrieve column names & types of a table
+            /// </summary>
+            public static string GetColumnDefinitionsQuery(string schema, string tableName)
+            {
+                return $@"
+                    select	
+	                    {ColumnName}, DATA_TYPE +
+		                    case 
+			                    when CHARACTER_MAXIMUM_LENGTH = -1 then '(max)'
+			                    when CHARACTER_MAXIMUM_LENGTH <> -1 then '(' + cast(CHARACTER_MAXIMUM_LENGTH as varchar(4)) + ')'
+                                when DATETIME_PRECISION is not null then '(' + cast(DATETIME_PRECISION as varchar(1)) + ')'
+			                    when DATA_TYPE in ('decimal', 'numeric') then '(' + cast(NUMERIC_PRECISION as varchar(9)) + ',' + + cast(NUMERIC_SCALE as varchar(9)) + ')'
+			                    else ''
+		                    end as {ColumnDefinition}	
+                    from 
+	                    INFORMATION_SCHEMA.COLUMNS c
+                    where
+	                    c.TABLE_NAME = '{tableName}'         
+                    and
+                        c.TABLE_SCHEMA = {schema}";
+            }
```

Comment:
Have we done a scan of this code to check for possible SQL injection vulnerabilities? There is a lot of manual query building that is making me a bit nervous.

Reply:
I agree, the manual query generation is definitely not ideal. Table name and schema come from the function `Attribute,` which the Function writer determines, so I think that is trustworthy. However, the data portion is generated on the fly, and could potentially come from user input. I'll double check if `OPENJSON` is vulnerable to injection.

Reply:
Correct, instead of using string interpolation, parameters should be used. I recommend using Dapper again as it will simplify code a lot, but not sure if you have any rule regarding taking external dependencies.